src/hotspot/share/cds/archiveHeapWriter.cpp

```diff
@@ -254,36 +252,46 @@ static int oop_sorting_rank(oop o) {
 // - objects that have only native pointers
 // - objects that have both native and oop pointers
 // - objects that have only oop pointers
-int ArchiveHeapWriter::compare_objs_by_oop_fields(int* a, int* b) {
-  oop oa = _source_objs->at(*a);
-  oop ob = _source_objs->at(*b);
-
-  int rank_a = oop_sorting_rank(oa);
-  int rank_b = oop_sorting_rank(ob);
+int ArchiveHeapWriter::compare_objs_by_oop_fields(HeapObjOrder* a, HeapObjOrder* b) {
+  int rank_a = a->_rank;
+  int rank_b = b->_rank;
 
   if (rank_a != rank_b) {
     return rank_a - rank_b;
   } else {
     // If they are the same rank, sort them by their position in the _source_objs array
-    return *a - *b;
+    return a->_index - b->_index;
   }
 }
 
 void ArchiveHeapWriter::sort_source_objs() {
+  log_info(cds)("sorting heap objects");
+  int len = _source_objs->length();
+  _source_objs_order = new GrowableArrayCHeap<HeapObjOrder, mtClassShared>(len);
+
+  for (int i = 0; i < len; i++) {
+    oop o = _source_objs->at(i);
+    int rank = oop_sorting_rank(o);
+    HeapObjOrder os = {i, rank};
+    _source_objs_order->append(os);
+  }
+  log_info(cds)("computed ranks");
   _source_objs_order->sort(compare_objs_by_oop_fields);
+  log_info(cds)("sorting heap objects done");
 }
 
 void ArchiveHeapWriter::copy_source_objs_to_buffer(GrowableArrayCHeap<oop, mtClassShared>* roots) {
   sort_source_objs();
   for (int i = 0; i < _source_objs_order->length(); i++) {
-    int src_obj_index = _source_objs_order->at(i);
+    int src_obj_index = _source_objs_order->at(i)._index;
     oop src_obj = _source_objs->at(src_obj_index);
     HeapShared::CachedOopInfo* info = HeapShared::archived_object_cache()->get(src_obj);
     assert(info != nullptr, "must be");
     size_t buffer_offset = copy_one_source_obj_to_buffer(src_obj);
     info->set_buffer_offset(buffer_offset);
 
     _buffer_offset_to_source_obj_table->put(buffer_offset, src_obj);
+    _buffer_offset_to_source_obj_table->maybe_grow();
```

Comment:
Similar to Calvin's comment, is it better to call `put_if_absent()` and then `maybe_grow()` if a new entry is created?

Reply:
Changed to `put_when_absent()` as this entry is guarantee to not exist.